--- audited-reference/contracts/RainbowRouter.sol	2026-01-16 19:30:00.814303600 -0600
+++ contracts/OkuRouter.sol	2026-01-16 19:29:48.504146800 -0600
@@ -1,69 +1,16 @@
-/*
-
-                                                            ▐██▒           ███
-                                            ,╓, ▄▄  ,▄▄▄,   .▄▄. ,╓, ▄▄▄   ██▌ ▄▄▄     ,▄▄▄,  ,╓╓   ╓╓   ,╓
-                                            ██████ ███▀▀██⌐ ▐██⌐ ███▀▀███⌐ ████▀▀███ ╓██▀▀▀██▄`██▌ ▐██▌ ▐██
-                                            ███    ▄▄█████▌ ▐██⌐ ██▌  ▐██▌ ██▌   ║██⌐███   ▓██ ╙██▄█▌██▄██⌐
-                                            ██▌   ▐██▄▄███▌,▐██⌐ ██▌  ▐██▌ ███▓▄▄███ ╙██▄▄▄██▀  ║███¬╙███▌
-                                            ╙╙└    ╙╙▀╙─╙▀▀└"╙╙` ╙╙└   ╙╙" ╙╙`╙╙▀▀└    ╙╙▀▀╙`    ╙╙└  ╙╙╙
-
-
-
-                                            _," _   _"""ⁿ=-,  _
-                                            ⌠            _    __"=.__
-                                            ▐░......   _  _          "=._
-                                            ▐░░░░░░░░░░░░░░░░.           "= _
-                                            ╚╩╩╩╩╩╩δφφφφ░░░░░░░░░░░          >__
-                                            ▐░░░░░░░░__ _ ╙╙╚╩φφ░░░░░░░        ^=_
-                                            ▐░░░░░░░░░░░░░░░,░  `╙╠φ░░░░░░░       ⁿ
-                                            ▐░░░░░░░░░░░░░░░░░░░░._ `╚Åφ░░░░░       " _
-                                            ╚╠╠╠╠╠╠╠╬╬╬▒▒φ░░░░░░░░░░░░  ╙╠░░░░░       "
-                                            ╚╝╝╝╝╝╝╝╬╬╬╠╢╬╠╬╠╬▒░░░░░░░░░░ "╚φ░░░░       ½_
-                                            ▐░░░░░░░░░░;░╙╙╝╬╠╬╠╠▒▒░░░░░░░░_ ╚φ░░░░      "_
-                                            ╚▒φφφ░░░░░░░░░░░░-╙╚╬╠╠╠╬▒░░░░░░░ `╠▒░░░░      ,
-                                            ╞╬╬╬╠╠╠╬╬╬╬▒φ▒░░░░░░░╙╚╬╬╠╬▒▒░░░░░ .╙╠░░░░      ≥
-                                            _▒░░░ΓΓ╙╙╙╚╩╬╠╬╠▒▒░░░░░░╙╬╬╠╬▒▒░░░░░' ╠▒░░░░     ≥
-                                            `╙ⁿⁿ≈≈σ╓░ '╙╙╚╬╠╬▒░░░░░░╙╬╬╠╬▒░░░░░  ╠▒░░░░     [
-                                                        _╙Θ░ ░╙╠╠╬╬▒░░░░░╬╠╠╠▒▒░░░░  ╠▒░░░░    '_
-                                                        _╙φ░'╙╠╠╬▒░░░░░╟╠╠╠▒░░░░░ _╠▒░░░     ░_
-                                                            _`φ ░╚╬╠╠▒░░░░║╠╠╠▒░░░░░.`╠░░░░     [
-                                                            _╚░⌡╚╠╬╬▒░░░░╠╠╠╬▒░░░░░ ╠▒░░░░    ░
-                                                                _╙░⌡╚╠╠╬▒░░░"▒╠╠╬▒░░░░ ⌠╠░░░░    ⌡_
-                                                                ╠ ░╠╠╠╬▒░░░║╠╬╠╬▒░░░ _╠▒░░░     Γ
-                                                                ▐░░░╠╠╠▒░░░╟╠╠╠╬▒░░░░ ╠▒░░░░    [
-                                                                _░.░╠╠╠▒░░░▐╬╠╠╬▒░░░░[╠╬░░░░    │
-                                                                _╙φ░╠╠╠╬▒░░▐╬╬╠╬╬▒░░░[╠╬░░░░  ░≥_
-                                                                    ____ ____  __ _______ ____
-
-
-    /$$$$$$                                           /$$$$$$                                                                /$$
-    /$$__  $$                                         /$$__  $$                                                              | $$
-    | $$  \__/ /$$  /$$  /$$  /$$$$$$   /$$$$$$       | $$  \ $$  /$$$$$$   /$$$$$$   /$$$$$$   /$$$$$$   /$$$$$$   /$$$$$$  /$$$$$$    /$$$$$$   /$$$$$$
-    |  $$$$$$ | $$ | $$ | $$ |____  $$ /$$__  $$      | $$$$$$$$ /$$__  $$ /$$__  $$ /$$__  $$ /$$__  $$ /$$__  $$ |____  $$|_  $$_/   /$$__  $$ /$$__  $$
-    \____  $$| $$ | $$ | $$  /$$$$$$$| $$  \ $$      | $$__  $$| $$  \ $$| $$  \ $$| $$  \__/| $$$$$$$$| $$  \ $$  /$$$$$$$  | $$    | $$  \ $$| $$  \__/
-    /$$  \ $$| $$ | $$ | $$ /$$__  $$| $$  | $$      | $$  | $$| $$  | $$| $$  | $$| $$      | $$_____/| $$  | $$ /$$__  $$  | $$ /$$| $$  | $$| $$
-    |  $$$$$$/|  $$$$$/$$$$/|  $$$$$$$| $$$$$$$/      | $$  | $$|  $$$$$$$|  $$$$$$$| $$      |  $$$$$$$|  $$$$$$$|  $$$$$$$  |  $$$$/|  $$$$$$/| $$
-    \______/  \_____/\___/  \_______/| $$____/       |__/  |__/ \____  $$ \____  $$|__/       \_______/ \____  $$ \_______/   \___/   \______/ |__/
-                                    | $$                       /$$  \ $$ /$$  \ $$                     /$$  \ $$
-                                    | $$                      |  $$$$$$/|  $$$$$$/                    |  $$$$$$/
-                                    |__/                       \______/  \______/                      \______/
-
-
-*/
-
 //SPDX-License-Identifier: GPL-3.0
-pragma solidity =0.8.11;
-import "@rari-capital/solmate/src/utils/SafeTransferLib.sol";
-import "@rari-capital/solmate/src/tokens/ERC20.sol";
+pragma solidity ^0.8.27;
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
 import "./routers/BaseAggregator.sol";
+import "./libraries/SafeTransferLib.sol";
 
-/// @title Rainbow swap aggregator contract
-contract RainbowRouter is BaseAggregator {
-    /// @dev The address that is the current owner of this contract
-    address public owner;
 
-    /// @dev Event emitted when the owner changes
-    event OwnerChanged(address indexed newOwner, address indexed oldOwner);
+/// @title Oku swap aggregator contract
+contract OkuRouter is BaseAggregator, Ownable {
+    string public name;
+    string public version;
 
     /// @dev Event emitted when a swap target gets added
     event SwapTargetAdded(address indexed target);
@@ -81,15 +28,22 @@
     /// @dev Event emitted when ETH fees are withdrawn
     event EthWithdrawn(address indexed target, uint256 amount);
 
-    /// @dev modifier that ensures only the owner is allowed to call a specific method
-    modifier onlyOwner() {
-        require(msg.sender == owner, "ONLY_OWNER");
-        _;
-    }
+    /// @dev Event emitted when a valid signer gets added
+    event ValidSignerAdded(address indexed target);
+
+    /// @dev Event emitted when a valid signer gets removed
+    event ValidSignerRemoved(address indexed target);
 
-    constructor() {
-        owner = msg.sender;
+    /// @dev Event emitted when the contract is paused
+    event ContractPaused(address indexed account);
+
+    /// @dev Event emitted when the contract is unpaused
+    event ContractUnpaused(address indexed account);
+
+    constructor(string memory _name, string memory _version) BaseAggregator(_name, _version) Ownable(msg.sender) {
         status = 1;
+        name = _name;
+        version = _version;
     }
 
     /// @dev We don't want to accept any ETH, except refunds from aggregators
@@ -97,7 +51,7 @@
     /// This is done by evaluating the value of status, which is set to 2
     /// only during swaps due to the "nonReentrant" modifier
     receive() external payable {
-        require(status == 2 || msg.sender == owner, "NO_RECEIVE");
+        require(status == 2 || msg.sender == owner(), "NO_RECEIVE");
     }
 
     /// @dev method to add or remove swap targets from swapTargets
@@ -114,6 +68,36 @@
         }
     }
 
+    /// @dev method to add or remove valid signers from validSigners
+    /// This is required so we only approve "trusted" signers
+    /// to sign transactions for this contrract
+    /// @param target address of the signer to add
+    /// @param add flag to add or remove the swap target
+    function updateValidSigner(address target, bool add) external onlyOwner {
+        validSigners[target] = add;
+        if (add) {
+            emit ValidSignerAdded(target);
+        } else {
+            emit ValidSignerRemoved(target);
+        }
+    }
+
+    /// @dev Pauses all swap operations on the contract
+    /// Only the owner can call this function
+    /// Emits a Paused event from the Pausable contract
+    function pause() external onlyOwner {
+        _pause();
+        emit ContractPaused(msg.sender);
+    }
+
+    /// @dev Unpauses all swap operations on the contract
+    /// Only the owner can call this function
+    /// Emits an Unpaused event from the Pausable contract
+    function unpause() external onlyOwner {
+        _unpause();
+        emit ContractUnpaused(msg.sender);
+    }
+
     /// @dev method to withdraw ERC20 tokens (from the fees)
     /// @param token address of the token to withdraw
     /// @param to address that's receiving the tokens
@@ -124,7 +108,7 @@
         uint256 amount
     ) external onlyOwner {
         require(to != address(0), "ZERO_ADDRESS");
-        SafeTransferLib.safeTransfer(ERC20(token), to, amount);
+        SafeERC20.safeTransfer(IERC20(token), to, amount);
         emit TokenWithdrawn(token, to, amount);
     }
 
@@ -136,15 +120,4 @@
         SafeTransferLib.safeTransferETH(to, amount);
         emit EthWithdrawn(to, amount);
     }
-
-    /// @dev Transfers ownership of the contract to a new account (`newOwner`).
-    /// @param newOwner address of the new owner
-    /// Can only be called by the current owner.
-    function transferOwnership(address newOwner) external virtual onlyOwner {
-        require(newOwner != address(0), "ZERO_ADDRESS");
-        require(newOwner != owner, "SAME_OWNER");
-        address previousOwner = owner;
-        owner = newOwner;
-        emit OwnerChanged(newOwner, previousOwner);
-    }
 }
